clippy_lints/src/one_unsafe_op_per_block.rs

```diff
@@ -0,0 +1,252 @@
+    /// ///
+    /// /// # Safety
+    /// ///
+    /// /// - `ptr` must be 4-byte aligned, point to four consecutive
+    /// ///   initialized bytes, and be valid for reads of 4 bytes.
+    /// /// - The bytes pointed to by `ptr` must represent a valid
+    /// ///   `char` when interpreted in the native byte order.
+    /// fn read_char(ptr: *const u8) -> char {
+    ///     // SAFETY: `ptr` is 4-byte aligned, points to four consecutive
+    ///     // initialized bytes, and is valid for reads of 4 bytes.
+    ///     let int_value = unsafe { *ptr.cast::<u32>() };
+    ///
+    ///     // SAFETY: The caller has guaranteed that the four bytes
+    ///     // pointed to by `bytes` represent a valid `char`.
+    ///     unsafe { char::from_u32_unchecked(int_value) }
+    /// }
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub ONE_UNSAFE_OP_PER_BLOCK,
+    restriction,
+    "more than one unsafe operation per `unsafe` block"
+}
+declare_lint_pass!(OneUnsafeOpPerBlock => [ONE_UNSAFE_OP_PER_BLOCK]);
+
+impl<'tcx> LateLintPass<'tcx> for OneUnsafeOpPerBlock {
+    fn check_block(&mut self, cx: &LateContext<'tcx>, block: &hir::Block<'_>) {
+        let mut unsafe_ops = vec![];
+        collect_block(cx, block, &mut unsafe_ops);
+        if unsafe_ops.len() > 1 {
+            span_lint_and_then(
+                cx,
+                ONE_UNSAFE_OP_PER_BLOCK,
+                block.span,
+                &format!(
+                    "this `unsafe` block contains {} unsafe operations, expected only one",
+                    unsafe_ops.len()
+                ),
+                |diag| {
+                    for (msg, span) in unsafe_ops {
+                        diag.span_note(span, msg);
+                    }
+                },
+            );
+        }
+    }
+}
+
+fn collect_block(cx: &LateContext<'_>, block: &hir::Block<'_>, unsafe_ops: &mut Vec<(&'static str, Span)>) {
+    if let BlockCheckMode::UnsafeBlock(_) = block.rules {
+        if let Some(expr) = block.expr {
+            collect_unsafe_exprs(cx, expr, unsafe_ops);
+        }
+        for stmt in block.stmts {
+            match stmt.kind {
+                StmtKind::Local(Local { init, els: r#else, .. }) => {
+                    if let Some(init) = init {
+                        collect_unsafe_exprs(cx, init, unsafe_ops);
+                    }
+                    if let Some(r#else) = r#else {
+                        collect_block(cx, r#else, unsafe_ops);
+                    }
+                },
+                StmtKind::Expr(e) | StmtKind::Semi(e) => collect_unsafe_exprs(cx, e, unsafe_ops),
+                StmtKind::Item(_) => {},
+            }
+        }
+    }
+}
+
+// This whole function is practically one big `match` block
+#[expect(
+    clippy::too_many_lines,
+    reason = "long match means long function, not much you can do here"
+)]
+fn collect_unsafe_exprs(cx: &LateContext<'_>, expr: &hir::Expr<'_>, unsafe_ops: &mut Vec<(&'static str, Span)>) {
```

Comment:
This could use [`for_each_expr_with_closures`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/visitors/fn.for_each_expr_with_closures.html) (or a struct that impls [`Visitor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html)) rather than calling itself recursively to be sure we're catching all the expressions